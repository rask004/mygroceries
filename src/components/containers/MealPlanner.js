import React, { Component } from 'react';
import * as DateFns from 'date-fns';
import { connect } from 'react-redux';
import {withRouter} from 'react-router-dom';
import PropTypes from 'prop-types';
import { addPlannedMeal, removePlannedMeal } from '../../store/actions';
import Card from '@material-ui/core/Card';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import {FaTrash, FaPlus} from 'react-icons/fa'
import 'typeface-roboto';
import '../../css/mealplanner.css';


class MealPlanner extends Component {
    constructor(props) {
        super(props);

        const now = new Date();

        this.state = {
            date: DateFns.format(now, "YYYY-MM-DD"),
            time: DateFns.format(now, "HH:mm"),
            recipeId: this.props.recipes[0].id,
        }

        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleChange = this.handleChange.bind(this);
    }

    handleSubmit(event) {
        event.preventDefault();
        
        const date = this.state.date;
        const time = this.state.time;
        const datetime = date + "T" + time;
        this.props.onAddMealplan(datetime, this.state.recipeId);
    };

    handleChange(event) {
        const name = event.target.name;
        const value = event.target.value;
        this.setState({[name]: value});
    };


    render() {
        const orderedMealplans = this.props.mealplans.slice().sort(
            (a,b) => DateFns.isBefore(DateFns.parse(a), DateFns.parse(b))
        );

        const renderMealplans = orderedMealplans.map( (item, step) => {
            const recipe = this.props.recipes.find(x => x.id === Number(item.recipeId));
            const title = recipe ? recipe.title : "";
            const dt = DateFns.parse(item.datetime);
            return (
                <li className="mealplan-item" key={step}>
                    <Grid container>
                        <Grid item xs={12} sm={6} md={3} className="content-date">
                            <Typography>{DateFns.format(dt, "YYYY-MM-DD")}</Typography>
                        </Grid>
                        <Grid item xs={12} sm={6} md={2} className="content-time">
                            <Typography>{DateFns.format(dt, "HH:mm")}</Typography>
                        </Grid>
                        <Grid item xs={12} sm={8} md={6} className="content-recipe">
                            <Typography>{title}</Typography>
                        </Grid>
                        <Grid item xs={12} sm={4} md={1}>
                            <Button className="action-remove" 
                                type="button" color="primary" size="medium" variant="contained"
                                onClick={() => this.props.onRemoveMealplan(item.datetime)}>
                                <FaTrash/>
                            </Button>
                        </Grid>
                    </Grid>
                </li>
            );
        });

        const recipeOptions = this.props.recipes.map( (item) => {
            return (
                <option key={item.id} value={item.id}>{item.title}</option>
            )
        });

        return (
            <Card className="mealplan">

                <Typography className="mealplan-title" variant="display1" 
                    gutterBottom={true}>Meal Planner</Typography>

                <ul>
                    {renderMealplans}
                </ul>

                <form className="add-form" onSubmit={this.handleSubmit}>
                    <Typography variant="title">Add Meal</Typography>
                    <Grid container>
                        <Grid item xs={12} sm={6} md={3}>
                            <TextField
                                name="date"
                                type="date"
                                className="picker-date"
                                value={this.state.date}
                                onChange={this.handleChange}
                                inputProps={{
                                    min: DateFns.format(Date(), "YYYY-MM-DD")
                                }}
                            />
                        </Grid>
                        <Grid item xs={12} sm={6} md={2}>
                            <TextField
                                name="time"
                                type="time"
                                className="picker-time"
                                value={this.state.time}
                                onChange={this.handleChange}
                                inputProps={{
                                    min: DateFns.format(Date(), "HH:mm"),
                                }}
                            />
                        </Grid>
                        <Grid item xs={12} sm={8} md={6}>
                            <FormControl>
                                <Select
                                    className="select-recipe"
                                    value={this.state.recipeId}
                                    onChange={this.handleChange}
                                    inputProps={{
                                        name: 'recipeId'    
                                    }}
                                >
                                    {recipeOptions}
                                </Select>
                            </FormControl>
                        </Grid>
                        <Grid item xs={12} sm={4} md={1}>
                            <Button className="action-add" 
                                type="submit" color="primary" size="medium" variant="contained">
                                <FaPlus/>
                            </Button>
                        </Grid>
                    </Grid>
                </form>                
            </Card>
        );
    }
}

MealPlanner.propTypes = {
    mealplans: PropTypes.array.isRequired,
    recipes: PropTypes.array.isRequired,
    onAddMealplan: PropTypes.func.isRequired,
    onRemoveMealplan: PropTypes.func.isRequired,
}


const mapStateToProps = state => {
	return {
        mealplans: state.mealplans,
        recipes: state.recipes,
    }
}

const mapDispatchToProps = dispatch => {
	return {
		onAddMealplan(datetime, recipeId) {
			dispatch(
				addPlannedMeal(datetime, recipeId)
			);
        },
        onRemoveMealplan(datetime) {
            dispatch(
                removePlannedMeal(datetime)
            );
        }
	}
}


export {MealPlanner};

const ConnectedMealPlanner = connect(mapStateToProps, mapDispatchToProps)(MealPlanner);
export default withRouter(ConnectedMealPlanner);