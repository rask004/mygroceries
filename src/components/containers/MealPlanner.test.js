import React from 'react';
import {create} from 'react-test-renderer';
import * as D from 'date-fns';

import {MealPlanner} from './MealPlanner';

import sampleData from '../../data/initialState.json';


const mealplans = sampleData.mealplans;
const recipes = sampleData.recipes;

const mealplanner = <MealPlanner
        mealplans={mealplans}
        recipes={recipes}
        onAddMealplan={jest.fn()}
        onRemoveMealplan={jest.fn()}
    />  


describe('MealPlanner Component rendering tests', () => {
    test('rendering snapshot', () => {
        const component = create(mealplanner);
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('datetimes and details are rendered correctly', () => {
        const component = create(mealplanner).root;
        const listitems = component.findAllByType('li');
        const orderedMealplans = mealplans.slice().sort(
            (a,b) => D.isBefore(D.parse(a), D.parse(b))
        );
        listitems.forEach( (item, step) => {
            const expectedMealplan = orderedMealplans[step];
            // immediate childs are within Typography components (material ui).
            const dateItem = item.findByProps({className:'content-date'}).props.children.props.children;
            const timeItem = item.findByProps({className:'content-time'}).props.children.props.children;
            const recipeItem = item.findByProps({className:'content-recipe'}).props.children.props.children;

            const recipe = recipes.find(x => x.title === recipeItem);
            const recipeId = recipe.id;

            const actualMealPlan = {
                datetime: dateItem + "T" + timeItem,
                recipeId: recipeId
            };

            expect(actualMealPlan).toEqual(expectedMealplan);
        });
    });
});

describe('MealPlanner Component behaviour tests', () => {

    test('click add mealplan calls onAddMealplan', () => {
        const component = create(mealplanner).root;

        const expectedDatetime = "2019-02-27T07:00";
        const expectedRecipeId = recipes[1].id;

        const mealPlanForm = component.findByProps({className:"add-form"});
        const dateField = mealPlanForm.findByProps({className:'picker-date'});
        const timeField = mealPlanForm.findByProps({className:'picker-time'});
        const recipeField = mealPlanForm.findByProps({className:'select-recipe'});
        dateField.props.onChange({
            target: {
                name:  "date",
                value: "2019-02-27"
            }
        });
        timeField.props.onChange({
            target: {
                name:  "time",
                value: "07:00"
            }
        });
        recipeField.props.onChange({
            target: {
                name:  "recipeId",
                value: recipes[1].id
            }
        });
        mealPlanForm.props.onSubmit({ preventDefault: () => {}});

        expect(component.props.onAddMealplan).toHaveBeenCalledWith(expectedDatetime, expectedRecipeId);
    });

    test('click remove mealplan calls onRemoveMealplan', () => {
        const component = create(mealplanner).root;
        const orderedMealplans = mealplans.slice().sort(
            (a,b) => D.isBefore(D.parse(a), D.parse(b))
        );
        const targetDateTime = orderedMealplans[0].datetime;

        const listItem = component.findAllByType('li')[0];
        const removeButton = listItem.findByProps({className:'action-remove'});
        removeButton.props.onClick();

        expect(component.props.onRemoveMealplan).toHaveBeenCalledWith(targetDateTime);
    });
});
