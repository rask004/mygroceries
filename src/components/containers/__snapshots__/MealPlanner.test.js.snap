// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealPlanner Component rendering tests rendering snapshot 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1 mealplan"
>
  <h1
    className="MuiTypography-root-29 MuiTypography-display1-33 MuiTypography-gutterBottom-56 mealplan-title"
  >
    Meal Planner
  </h1>
  <ul>
    <li
      className="mealplan-item"
    >
      <div
        className="MuiGrid-container-65"
      >
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-3-124 content-date"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            2018-07-04
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 content-time"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            19:30
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-8-115 MuiGrid-grid-md-6-127 content-recipe"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            test recipe
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-4-111 MuiGrid-grid-md-1-122"
        >
          <button
            className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-163"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-245"
            />
          </button>
        </div>
      </div>
    </li>
    <li
      className="mealplan-item"
    >
      <div
        className="MuiGrid-container-65"
      >
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-3-124 content-date"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            2018-07-05
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 content-time"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            00:00
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-8-115 MuiGrid-grid-md-6-127 content-recipe"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            another test recipe
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-4-111 MuiGrid-grid-md-1-122"
        >
          <button
            className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-163"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-245"
            />
          </button>
        </div>
      </div>
    </li>
    <li
      className="mealplan-item"
    >
      <div
        className="MuiGrid-container-65"
      >
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-3-124 content-date"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            2018-07-05
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123 content-time"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            07:15
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-8-115 MuiGrid-grid-md-6-127 content-recipe"
        >
          <p
            className="MuiTypography-root-29 MuiTypography-body1-38"
          >
            the third test recipe
          </p>
        </div>
        <div
          className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-4-111 MuiGrid-grid-md-1-122"
        >
          <button
            className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-remove"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-163"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-245"
            />
          </button>
        </div>
      </div>
    </li>
  </ul>
  <form
    className="add-form"
    onSubmit={[Function]}
  >
    <h2
      className="MuiTypography-root-29 MuiTypography-title-35"
    >
      Add Meal
    </h2>
    <div
      className="MuiGrid-container-65"
    >
      <div
        className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-3-124"
      >
        <div
          className="MuiFormControl-root-191 picker-date"
        >
          <div
            className="MuiInputBase-root-208 MuiInput-root-195 MuiInput-underline-199 MuiInputBase-formControl-209 MuiInput-formControl-196"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-218 MuiInput-input-203 MuiInputBase-inputType-221 MuiInput-inputType-206"
              disabled={false}
              min="2019-05-15"
              name="date"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="date"
              value="2019-05-15"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-6-113 MuiGrid-grid-md-2-123"
      >
        <div
          className="MuiFormControl-root-191 picker-time"
        >
          <div
            className="MuiInputBase-root-208 MuiInput-root-195 MuiInput-underline-199 MuiInputBase-formControl-209 MuiInput-formControl-196"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-218 MuiInput-input-203 MuiInputBase-inputType-221 MuiInput-inputType-206"
              disabled={false}
              min="16:00"
              name="time"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="time"
              value="16:48"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-8-115 MuiGrid-grid-md-6-127"
      >
        <div
          className="MuiFormControl-root-191"
        >
          <div
            className="MuiInputBase-root-208 MuiInput-root-195 MuiInput-underline-199 MuiInputBase-formControl-209 MuiInput-formControl-196 select-recipe"
            onClick={[Function]}
          >
            <div
              className="MuiSelect-root-225"
            >
              <div
                aria-haspopup="true"
                aria-pressed="false"
                className="MuiSelect-select-226 MuiSelect-selectMenu-229 MuiInputBase-input-218 MuiInput-input-203"
                id="select-recipeId"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
              >
                test recipe
              </div>
              <input
                name="recipeId"
                type="hidden"
                value={1}
              />
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-232 MuiSelect-icon-231"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-item-66 MuiGrid-grid-xs-12-105 MuiGrid-grid-sm-4-111 MuiGrid-grid-md-1-122"
      >
        <button
          className="MuiButtonBase-root-188 MuiButton-root-162 MuiButton-contained-173 MuiButton-containedPrimary-174 MuiButton-raised-176 MuiButton-raisedPrimary-177 action-add"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="submit"
        >
          <span
            className="MuiButton-label-163"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-245"
          />
        </button>
      </div>
    </div>
  </form>
</div>
`;
